thirdparty/lpeg/CMakeLists.txt

```diff
@@ -9,9 +9,10 @@ enable_language(C)
 assert_var_defined(CC)
 assert_var_defined(DYNLIB_CFLAGS)
 assert_var_defined(LUA_DIR)
-assert_var_defined(LUAJIT_LIB)
 
-set(BUILD_CMD_STR "${CC} ${DYNLIB_CFLAGS} -I${LUA_DIR}/src ${LUAJIT_LIB} -o lpeg.so lpcap.c lpcode.c lpprint.c lptree.c lpvm.c")
+# assuming lpeg.so is installed under $(OUTPUT_DIR)/rocks/lib/lua/5.1 and
+# libluajit.so is in $(OUTPUT_DIR)/libs
+set(BUILD_CMD_STR "${CC} ${DYNLIB_CFLAGS} -I${LUA_DIR}/src -Wl,-rpath,'../../../../libs' -lluajit -o lpeg.so lpcap.c lpcode.c lpprint.c lptree.c lpvm.c")
```

Comment:
The `$ORIGIN` path directs `ld` to look up linked shared objects just in the directory of the original binary file. In our case, `ld` will look for `libz` and `libjpeg` in the same directory as `libmupdf` and it will find them in the `libs`.

Aha, I see. We are talking about the Android platform. I can see that the `lpeg.so` is in `/data/data/org.koreader.launcher/files/rocks/lib/lua/5.1/lpeg.so` from which `../../../../libs` is  just `/data/data/org.koreader.launcher/files/libs` and the `libluajit.so` is right there.

But the recursive dependency loading function of shared libraries does differ a lot with `ld` on other Linux distributions. @hwhw implemented it here[https://github.com/koreader/android-luajit-launcher/blob/master/assets/dl.lua#L73]. From what I understand it doesn't respect the `rpath` at all. All it needs for a shared library are the dependent library names. It will look up the dependent libraries at [predefined paths](https://github.com/koreader/android-luajit-launcher/blob/master/assets/android.lua#L1350-L1353). And if an `so` file is found it will use `bionic`(Android's glibc)'s `dlopen` function to load it.

My explanation for your observation is that on your Android somehow hwhw's `dl` module cannot find `libluajit.so` while the `bionic` can find it with `rpath` in an extraordinary way.